src/fah/client/Units.cpp

```diff
@@ -143,3 +164,95 @@ void Units::load() {
 
   if (empty()) update();
 }
+
+state_ Units::processPossibility(int32_t cpus, std::set<std::string> gpus, unsigned i) {
+  state_ result;
+  result.choice = i;
+  unsigned n = size() - 1;
+
+  while(i != 0) {
+    // Check if workunit is selected.
+    bool selected = (i & 1) == 1 ? true : false;
+
+    auto &unit = *get(n).cast<Unit>();
+    auto &unitGPUs = *unit.getGPUs();
+    uint32_t unitCPUs = unit.getCPUs();
+
+    if(selected) {
+      if(result.largestCpuWu < unitCPUs) result.largestCpuWu = unitCPUs;
+      result.cpus += unit.getCPUs();
+      for (unsigned j = 0; j < unitGPUs.size(); j++) {
+        if(!gpus.count(unitGPUs.getString(j))) return result;
+        gpus.erase(unitGPUs.getString(j));
+        result.gpus++;
+      }
+    }
+    if(result.cpus > cpus) return result;
+
+    i >>= 1;
+    n--;
+  }
+  result.feasible = true;
+  return result;
+}
+
+bool Units::compare(state_ a, state_ b) {
+  if(a.gpus > b.gpus) return true;
+  if(a.cpus > b.cpus) return true;
+  if(a.units < b.units) return true;
+  if(a.units == b.units && a.largestCpuWu > b.largestCpuWu) return true;
+  return false;
+}
+
+
+state_ Units::removeUsedResources(int32_t cpus, std::set<std::string> &gpus) {
+
+  unsigned wus = size();
+  unsigned mask = generateMask(cpus, gpus);
+  state_ result;
+
+  for(unsigned i = 1; i < pow(2, wus) - 1; i++)
```

Comment:
Using a bit mask to track the possible combinations is not viable.  It limits us to either 32 or 64 GPUs, which we don't want to do.  It's also inefficient because you do not need to iterate all possible combinations.  You can keep track of sets of WUs using ``std::set<unsigned>``. You can do the search recursively like this:

```
typedef std::set<unsigned> wu_set_t;

bool isBetter(const wu_set_t &a, const wu_set_t &b);
bool usesTooManyResources(const wu_set_t &s);

wu_set_t findBest(const wu_set_t &current, int i) {
  wu_set_t best = current;

  for (int j = i; j < wus.size(); j++) {
    wu_set_t s = current;
    s.insert(j);
    if (usesTooManyResources(s)) continue; // Important
    s = findBest(s, j + 1);
    if (isBetter(s, best)) best = s;
  }

  return best;
}

wu_set_t best = findBest(wu_set_t(), 0);
```

This should be fast enough.  If it's not, there are further optimizations.